src/main/java/events/writer/PubSubEventWriter.java

```diff
@@ -0,0 +1,172 @@
+        } catch (IOException e) {
+            logger.error("Error creating pubsub events.publisher. Error: " + e.getMessage());
+        }
+    }
+
+    /**
+     *
+     * @param saPath
+     * @return
+     * @throws FileNotFoundException
+     */
+    private InputStream getCredentials(String saPath) throws FileNotFoundException {
+        File credentialsFile = new File(saPath);
+        return new FileInputStream(credentialsFile);
+    }
+
+    /**
+     *
+     * @param builder
+     * @param proxyHost
+     * @param proxyPort
+     * @throws NumberFormatException
+     */
+    private void configureChannelProxy(Publisher.Builder builder, String proxyHost, String proxyPort) throws NumberFormatException {
+        SocketAddress proxySocketAddress = new InetSocketAddress(proxyHost, Integer.parseInt(proxyPort));
+        builder.setChannelProvider(
+                InstantiatingGrpcChannelProvider.newBuilder()
+                        .setChannelConfigurator(managedChannelBuilder -> managedChannelBuilder.proxyDetector(new ProxyDetector() {
+                            @Nullable
+                            @Override
+                            public ProxiedSocketAddress proxyFor(SocketAddress socketAddress) {
+                                if (socketAddress == null) return null;
+
+                                return HttpConnectProxiedSocketAddress.newBuilder()
+                                        .setTargetAddress((InetSocketAddress) socketAddress)
+                                        .setProxyAddress(proxySocketAddress)
+                                        .build();
+                            }
+                        }))
+                        .build()
+        );
+    }
+
+    /**
+     *
+     * @param event
+     */
+    public void publishEvent(Event event) {
+        Gson gson = new Gson();
+
+        logger.info("Publishing event");
+        String stringEvent = gson.toJson(event);
+        ByteString data = ByteString.copyFromUtf8(stringEvent);
+        try {
+            PubsubMessage pubsubMessage = PubsubMessage.newBuilder()
+                    .setData(data)
+                    .build();
+            ApiFuture<String> future = this.publisher.publish(pubsubMessage);
+            ApiFutures.addCallback(
+                    future,
+                    new ApiFutureCallback<String>() {
+
+                        @Override
+                        public void onFailure(Throwable e) {
+                            logger.error("Error publishing message : " + e.getMessage());
+                        }
+
+                        @Override
+                        public void onSuccess(String messageId) {
+                            logger.info("Published message ID: " + messageId);
+                        }
+                    },
+                    MoreExecutors.directExecutor());
+            int retries = 0;
+            while (!future.isDone() && retries <= 10) {
```

Comment:
Can we use get() with a timeout for the future result?

Reply:
We can change the code, but we added the callback to check if the message was sent correctly

Reply:
I see that callback feature is recommended in place of `get()` so the thread is not blocked. Here however there is the block below with sleep() for checking if the future is done after the callback. Can that block be removed? It is best to avoid Thread.sleep() in the code.
```int retries = 0;
            while (!future.isDone() && retries <= 10) {
                logger.info("Future not done yet");
                retries++;
                Thread.sleep(300);
            }
```

Reply:
You are right, I'm going to change the behaviour not to block the code if the future is done.